src/inhomcorr/methods/unet.py

```diff
@@ -0,0 +1,228 @@
+        elif dim == 3:
+            interp_mode = 'trilinear'
+            conv_op = nn.Conv3d
+
+        self.interp_mode = interp_mode
+
+        padding = int(np.floor(kernel_size / 2))
+        self.upconvs = nn.ModuleList([conv_op(n_ch_list[i],
+                                              n_ch_list[i + 1],
+                                              kernel_size=kernel_size,
+                                              padding=padding,
+                                              bias=bias)
+                                      for i in range(len(n_ch_list) - 1)])
+        self.dec_blocks = nn.ModuleList([ConvBlock(dim,
+                                                   n_ch_list[i],
+                                                   n_ch_list[i + 1],
+                                                   n_convs_per_stage,
+                                                   kernel_size=kernel_size,
+                                                   bias=bias)
+                                         for i in range(len(n_ch_list) - 1)])
+
+    def forward(self, x, encoder_features):
+        """Forward pass of the decoder."""
+        for i in range(len(self.dec_blocks)):
+            # x        = self.upconvs[i](x)
+            enc_features = encoder_features[i]
+            enc_features_shape = enc_features.shape
+            x = nn.functional.interpolate(
+                x, enc_features_shape[2:], mode=self.interp_mode)
+            x = self.upconvs[i](x)
+            x = torch.cat([x, enc_features], dim=1)
+            x = self.dec_blocks[i](x)
+        return x
+
+
+class UNet(nn.Module):
+    """UNet model."""
+
+    def __init__(self, hparams: HyperParameters):
+        super(UNet, self).__init__()
+
+        self.encoder = Encoder(hparams.dim,
+                               hparams.nChIn,
+                               hparams.nEncStages,
+                               hparams.nConvsPerStage,
+                               hparams.nFilters,
+                               hparams.KernelSize,
+                               hparams.Bias)
+        self.decoder = Decoder(hparams.dim,
+                               hparams.nFilters *
+                               (2**(hparams.nEncStages - 1)),
+                               hparams.nEncStages,
+                               hparams.nConvsPerStage,
+                               hparams.nFilters * (hparams.nEncStages * 2),
+                               kernel_size=hparams.KernelSize,
+                               bias=hparams.Bias)
+
+        if hparams.dim == 1:
+            conv_op = nn.Conv1d
+        elif hparams.dim == 2:
+            conv_op = nn.Conv2d
+        elif hparams.dim == 3:
+            conv_op = nn.Conv3d
+
+        self.c1x1 = conv_op(hparams.nFilters,
+                            hparams.nChOut,
+                            kernel_size=1, padding=0, bias=hparams.Bias)
+        self.res_connection = hparams.ResConnection
```

Comment:
i would remove this residual connection from the parameters. this is not useful in the general case but can be used to the model using the unet if needed.